doc/pandas-2.0/source/goals.rst

```diff
@@ -0,0 +1,194 @@
+
+* The pandas 0.x series of releases have consisted with huge amounts of
+  iterative improvements to the library along with some major new features, bug
+  fixes, and improved documentation. There have also been a series of
+  deprecations, API changes, and other evolutions of pandas's API to account
+  for suboptimal design choices (for example: the ``.ix`` operator) made in the
+  early days of the project (2010 to 2012).
+* The unification of Series and DataFrame internals to be based on a common
+  ``NDFrame`` base class and "block manager" data structure (originally created
+  by me in 2011, and heroically driven forward to its modern form by Jeff
+  Reback), while introducing many benefits to pandas, has come to be viewed as
+  a long-term source of technical debt and code complexity.
+* pandas's ability to support an increasingly broad set of use cases has been
+  significantly constrained (as will be examined in detail in these documents)
+  by its tight coupling to NumPy and therefore subject to various limitations
+  in NumPy.
+* Making significant functional additions (particularly filling gaps in NumPy)
+  to pandas, particularly new data types, has grown increasingly complex with
+  very obvious accumulations of technical debt.
+* pandas is being used increasingly for very large datasets on machines with
+  many cores and large amounts of RAM (100s of gigabytes to terabytes). It
+  would be nice to be able to better utilize these larger, beefier systems
+  within a single Python process.
+* pandas is being used increasingly as a computational building block of some
+  larger system, such as Dask or Apache Spark. We should consider reducing the
+  overhead for making data accessible to pandas (i.e. via memory-mapping or
+  other low-overhead memory sharing).
+* Rough edges in pandas's implementation (e.g. its handling of missing data
+  across data types) are being exposed to users.
+
+These documents are largely concerned with pandas's internal design, which is
+mostly invisible to average users. Advanced users of pandas are generally
+familiar with some of these internal details, particular around performance and
+memory use, and so the degree to which users are impacted will vary quite a
+lot.
+
+Key areas of work
+=================
+
+Possible changes or improvements to pandas's internals fall into a number of
+different buckets to be explored in great detail:
+
+* **Decoupling from NumPy while preserving interoperability**: by eliminating
+  the presumption that pandas objects internally must contain data stored in
+  NumPy ``ndarray`` objects, we will be able to bring more consistency to
+  pandas's semantics and enable the core developers to extend pandas more
+  cleanly with new data types, data structures, and computational semantics.
+* **Exposing a pandas Cython and/or C/C++ API to other Python library
+  developers**: the internals of Series and DataFrame are only weakly
```

Comment:
I think it's prob worth some thought how to make a clean API between the storage back end, pandas structures and the dispatch (compute) machinery

eg say we wanted to swap in something like: https://github.com/alimanfoo/zarr

this actually in theory should be easy as it has numpy storage / indexing semantics 

in a similar way would be nice to define an API to a compute engine (like numexpr or numba)

mainly need a consistent way of handling dtype conversions (this of course is very similar to the problem arrow is designed to solve) except here we could use numpy as the intermediary